pkg/bgpv1/manager/reconcile.go

```diff
@@ -626,26 +626,33 @@ func (r *LBServiceReconciler) svcDiffReconciliation(ctx context.Context, sc *Ser
 // desired state.
 func (r *LBServiceReconciler) svcDesiredRoutes(newc *v2alpha1api.CiliumBGPVirtualRouter, svc *slim_corev1.Service, ls localServices) ([]netip.Prefix, error) {
 	if newc.ServiceSelector == nil {
-		// If the vRouter has no service selector, there are no desired routes
+		// If the vRouter has no service selector, there are no desired routes.
 		return nil, nil
 	}
 
-	// Ignore non-loadbalancer services
+	// Ignore non-loadbalancer services.
 	if svc.Spec.Type != slim_corev1.ServiceTypeLoadBalancer {
 		return nil, nil
 	}
 
+	// The vRouter has a service selector, so determine the desired routes.
 	svcSelector, err := slim_metav1.LabelSelectorAsSelector(newc.ServiceSelector)
 	if err != nil {
 		return nil, fmt.Errorf("labelSelectorAsSelector: %w", err)
 	}
 
-	// Ignore non matching services
+	// Ignore non matching services.
 	if !svcSelector.Matches(serviceLabelSet(svc)) {
 		return nil, nil
 	}
 
-	// Ignore externalTrafficPolicy == Local && no local endpoints
+	// Ignore service managed by an unsupported LB class.
+	if svc.Spec.LoadBalancerClass != nil && *svc.Spec.LoadBalancerClass != v2alpha1api.BGPLoadBalancerClass {
```

Comment:
Does this mean that BGP CP handles services where `.spec.loadBalancerClass` is not set as well? If so, wouldn't it interfere with the services that the cloud provider is supposed to be handling?

https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class

> By default, .spec.loadBalancerClass is not set and a LoadBalancer type of Service uses the cloud provider's default load balancer implementation if the cluster is configured with a cloud provider using the --cloud-provider component flag.

MetalLB handles only services whose spec.loadBalancerClass matches the given lb class. Have we considered an option like that?

https://github.com/metallb/metallb/blob/4672d27ba6168e833c33a4f3ace337c079de57b3/speaker/main.go#L83

> load balancer class. When enabled, metallb will handle only services whose spec.loadBalancerClass matches the given lb class

Reply:
> Does this mean that BGP CP handles services where .spec.loadBalancerClass is not set as well? If so, wouldn't it interfere with the services that the cloud provider is supposed to be handling?

Yes, I think that is the current behavior as well due to user friendliness. As LB IPAM & BGP announcements is mostly an on-prem feature, most likely there will be no cloud provider controller handling the services when it is enabled - and in that case making the Cilium the default load balancer implementation probably makes sense.

Reply:
Thank you for your comment!

> in that case making the Cilium the default load balancer implementation probably makes sense.

So What motivates users to set LoadBalancerClass in the on-prem environment where Cilium is the default load balancer implementation?

Reply:
One motivation is described in the [original issue](https://github.com/cilium/cilium/issues/26556) - apart from BGP announcements, Cilium now also supports L2 announcement of LB services, so it can be used to distinguish between the two.

Apart from that, on-prem clusters may still have some other on-prem load-balancer deployed (such as MetalLB), for example if they used it in the past and now want to migrate to Cilium.

Reply:
From the original issue
> The L2 announcement feature does not have a loadbalancerClass yet, so we should add one.

Are we going to add a loadbalancerClass for L2 mode in a separate PR? So that users can set it to the services which they expect to be handled by the L2 mode (not BGP mode).

Reply:
Good point @ysksuzuki , I think it makes sense to have a separate PR for that - but that means we should not close the original issue with this PR @danehans

Reply:
Thank you for your explanation. It's clear now.

Reply:
@rastislavs https://github.com/cilium/cilium/issues/26556 was reopened so L2 mode can be resolved.